File: src/main/java/com/dominionwars/engine/CardInstance.java
package com.dominionwars.engine;

import com.dominionwars.model.CardDef;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

/** 运行时卡牌实例 */
public class CardInstance {
    private static final AtomicInteger SEQ = new AtomicInteger(1);

    public final int uid = SEQ.getAndIncrement();
    public final CardDef def;
    public final int ownerIdx;

    // 随从/统领状态
    public int attack, health, maxHealth;
    public int attacksUsed = 0;
    public int attacksPerTurn;
    public boolean summonedThisTurn = false;
    public boolean shield;                  // 圣盾未消耗
    public Set<String> keywords = new LinkedHashSet<>();

    // 统领
    public boolean isLeaderEntity = false;  // 该实例以统领身份在场
    public int durability = 0;              // 非随从统领耐久

    // 吟唱
    public int chantRemaining = 0;

    // 惩罚牌/惩罚效果激活状态：仅当因对方惩罚抽入手牌时为 true
    public boolean punishActivated = false;

    public CardInstance(CardDef def, int ownerIdx) {
        this.def = def;
        this.ownerIdx = ownerIdx;
        this.attack = def.attack;
        this.health = def.health;
        this.maxHealth = def.health;
        this.attacksPerTurn = def.attacksPerTurn;
        this.keywords = new LinkedHashSet<>(def.keywords);
        this.shield = keywords.contains(CardDef.KW_SHIELD);
        this.durability = def.leader ? def.leaderDef.durability : 0;
    }

    /** 进入卡组/手牌前重置运行时状态，避免洗回后抽到 0 血或负血随从。 */
    public void resetRuntimeState() {
        this.attack = def.attack;
        this.health = def.health;
        this.maxHealth = def.health;
        this.attacksUsed = 0;
        this.attacksPerTurn = def.attacksPerTurn;
        this.summonedThisTurn = false;
        this.keywords = new LinkedHashSet<>(def.keywords);
        this.shield = keywords.contains(CardDef.KW_SHIELD);
        this.durability = def.leader ? def.leaderDef.durability : 0;
        this.chantRemaining = 0;
        this.punishActivated = false;
        this.isLeaderEntity = false;
    }

    public boolean isMinionOnField() { return def.isMinion() || (isLeaderEntity && def.isMinion()); }
    public boolean has(String kw) { return keywords.contains(kw); }
    public boolean canAttackNow() {
        if (!def.isMinion() || attack <= 0) return false;
        if (summonedThisTurn && !has(CardDef.KW_CHARGE)) return false;
        return attacksUsed < attacksPerTurn;
    }
    public boolean alive() {
        if (def.isMinion()) return health > 0;
        if (isLeaderEntity) return durability > 0 || def.leaderDef.durability == 0;
        return true;
    }

    public String shortName() { return def.name + "#" + uid; }
    @Override public String toString() {
        if (def.isMinion()) return def.name + "(" + attack + "/" + health + ")";
        return def.name;
    }
}

File: src/Engine/Model/CardInstance.cs
using System;
using System.Collections.Generic;

namespace DominionWars.Engine.Model
{

public sealed class CardInstance
{
    public CardInstance(long instanceId, int ownerPlayerIndex, CardDefinition definition)
    {
        if (instanceId <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(instanceId));
        }

        if (ownerPlayerIndex is < 0 or > 1)
        {
            throw new ArgumentOutOfRangeException(nameof(ownerPlayerIndex));
        }

        Definition = definition ?? throw new ArgumentNullException(nameof(definition));
        InstanceId = instanceId;
        OwnerPlayerIndex = ownerPlayerIndex;
        Attack = definition.Attack;
        Health = definition.Health;
        MaxHealth = definition.Health;
        Durability = definition.LeaderDurability;
        Keywords = new HashSet<string>(definition.Keywords, StringComparer.OrdinalIgnoreCase);
        Shield = Keywords.Contains("圣盾");
    }

    public long InstanceId { get; }
    public int OwnerPlayerIndex { get; }
    public CardDefinition Definition { get; }
    public int Attack { get; set; }
    public int Health { get; set; }
    public int MaxHealth { get; set; }
    public bool Shield { get; set; }
    public int AttacksUsed { get; set; }
    public bool SummonedThisTurn { get; set; }
    public bool PunishActivated { get; set; }
    public int ChantRemaining { get; set; }
    public int Durability { get; set; }
    /// <summary>Temporary controller for CONTROL. Ownership never changes.</summary>
    public int? ControlledByPlayerIndex { get; set; }
    public int ControlTurnsRemaining { get; set; }
    /// <summary>木阵营封印状态：失去特殊能力且不能攻击。</summary>
    public bool Sealed { get; set; }
    public ISet<string> Keywords { get; }

    public bool IsMinion => Definition.IsMinion;
    public bool IsLeader => Definition.IsLeader;
    public bool IsLeaderEntity { get; set; }
    public int ControllerPlayerIndex => ControlledByPlayerIndex ?? OwnerPlayerIndex;
    public bool IsAlive => Health > 0;

    public bool HasKeyword(string keyword)
    {
        // A growth-sealed unit keeps its identity for snapshots and victory
        // checks, but its special abilities are inactive until the seal is
        // removed.  Centralising this guard prevents taunt/ward/charge and
        // future keyword consumers from accidentally bypassing that rule.
        return !Sealed
            && !string.IsNullOrWhiteSpace(keyword)
            && Keywords.Contains(keyword);
    }

    public void ResetRuntimeState()
    {
        Attack = Definition.Attack;
        Health = Definition.Health;
        MaxHealth = Definition.Health;
        Shield = false;
        AttacksUsed = 0;
        SummonedThisTurn = false;
        PunishActivated = false;
        ChantRemaining = 0;
        Durability = Definition.LeaderDurability;
        ControlledByPlayerIndex = null;
        ControlTurnsRemaining = 0;
        Sealed = false;
        IsLeaderEntity = false;
        Keywords.Clear();
        foreach (var keyword in Definition.Keywords)
        {
            Keywords.Add(keyword);
        }

        Shield = Keywords.Contains("圣盾");
    }
}
}
